engines/lab/resource.cpp

```diff
@@ -0,0 +1,417 @@
+	curRoom->_view[SOUTH] = readView(dataFile);
+	curRoom->_view[EAST] = readView(dataFile);
+	curRoom->_view[WEST] = readView(dataFile);
+	curRoom->_rules = readRule(dataFile);
+
+	_vm->updateMusicAndEvents();
+	delete dataFile;
+}
+
+void Resource::freeViews(uint16 roomNum) {
+	if (!_vm->_rooms)
+		return;
+
+	for (uint16 i = 0; i < 4; i++)
+		freeView(_vm->_rooms[roomNum]._view[i]);
+
+	freeRule(_vm->_rooms[roomNum]._rules);
+}
+
+Common::String Resource::translateFileName(const Common::String filename) {
+	Common::String upperFilename = filename;
+	upperFilename.toUppercase();
+	Common::String fileNameStrFinal;
+
+	if (upperFilename.hasPrefix("P:") || upperFilename.hasPrefix("F:")) {
+		if (_vm->_isHiRes)
+			fileNameStrFinal = "GAME/SPICT/";
+		else
+			fileNameStrFinal = "GAME/PICT/";
+
+		if (_vm->getPlatform() == Common::kPlatformAmiga) {
+			if (upperFilename.hasPrefix("P:")) {
+				fileNameStrFinal = "PICT/";
+			} else {
+				fileNameStrFinal = "LABFONTS/";
+				upperFilename += "T";	// all the Amiga fonts have a ".FONT" suffix
+			}
+		}
+	} else if (upperFilename.hasPrefix("LAB:")) {
+		if (_vm->getPlatform() != Common::kPlatformAmiga)
+			fileNameStrFinal = "GAME/";
+	} else if (upperFilename.hasPrefix("MUSIC:")) {
+		if (_vm->getPlatform() != Common::kPlatformAmiga)
+			fileNameStrFinal = "GAME/MUSIC/";
+		else
+			fileNameStrFinal = "MUSIC/";
+	}
+
+	if (upperFilename.contains(':')) {
+		while (upperFilename[0] != ':') {
+			upperFilename.deleteChar(0);
+		}
+
+		upperFilename.deleteChar(0);
+	}
+
+	fileNameStrFinal += upperFilename;
+
+	return fileNameStrFinal;
+}
+
+Common::File *Resource::openDataFile(const Common::String fileName, uint32 fileHeader) {
+	Common::File *dataFile = new Common::File();
+	dataFile->open(translateFileName(fileName));
+	if (!dataFile->isOpen())
+		error("openDataFile: Couldn't open %s (%s)", translateFileName(fileName).c_str(), fileName.c_str());
+
+	if (fileHeader > 0) {
+		uint32 headerTag = dataFile->readUint32BE();
+		if (headerTag != fileHeader) {
+			dataFile->close();
+			error("openDataFile: Unexpected header in %s (%s) - expected: %d, got: %d", translateFileName(fileName).c_str(), fileName.c_str(), fileHeader, headerTag);
+		}
+	}
+
+	return dataFile;
+}
+
+Common::String Resource::readString(Common::File *file) {
+	byte size = file->readByte();
+	if (!size)
+		return Common::String("");
+
+	char *str = new char[size];
+	for (int i = 0; i < size; i++) {
+		char c = file->readByte();
+		// Decrypt char
+		c = (i < size - 1) ? c - 95 : '\0';
+		str[i] = c;
+	}
+
+	Common::String result = str;
+	delete[] str;
+	return result;
+}
+
+int16 *Resource::readConditions(Common::File *file) {
+	int16 i = 0, cond;
+	int16 *list = new int16[25];
+	memset(list, 0, 25 * sizeof(int16));
+
+	do {
+		cond = file->readUint16LE();
+		if (i < 25)
+			list[i++] = cond;
+	} while (cond);
+
+	return list;
+}
+
+RuleList *Resource::readRule(Common::File *file) {
+	char c;
+	RuleList *rules = new RuleList();
+
+	do {
```

Comment:
I think it would be simpler to write this with a single `while` loop without the need for any `if`:
```
while ((c = file->readSByte()) == 1) {
//  [...]
}
```